File: solutions/day02/solution.js
var maxBottlesDrunk = function(numBottles, numExchange) {
    let full = numBottles;
    let empty = 0;
    let ans = 0;
    let curEx = numExchange;

    while (full > 0) {
        ans += full;
        empty += full;
        full = 0;

        while (empty >= curEx) {
            empty -= curEx;
            full += 1;
            curEx += 1;
        }
    }
    return ans;
};

File: solutions/day02/solution.c
int maxBottlesDrunk(int numBottles, int numExchange) {
    int count = numBottles;
    int empty = numBottles;

    // Instead of simulation which takes n^2, we can just count of full bottles
    while (empty >= numExchange) {
        count++;
        // We subtract 1 from numExchange here because we are gaining 1 full bottles for numExchange empty bottles
        empty -= numExchange - 1;
        numExchange++;
    }

    return count;
}